PLR1711, RET501 should exempt annotated function where return type isn’t `-> None`

Ruff flags PLR1711 useless-return and RET501 unnecessary-return-none in this code:

```python
class BaseCache:
    def get(self, key: str) -> str | None:
        print(f"{key} not found")
        return None
```

```
test.py:4:9: RET501 [*] Do not explicitly `return None` in function if it is the only possible return value
test.py:4:9: PLR1711 [*] Useless `return` statement at end of function
```

But mypy complains if `return None` is removed:

```
test.py:2: error: Missing return statement  [return]
```

or if `return None` is changed to `return`:

```
test.py:4: error: Return value expected  [return-value]
```

mypy can be appeased by changing the return type annotation to `-> None`, and that’s okay in many cases (e.g. `Callable[..., None]` covariantly coerces to `Callable[..., str | None`), but it’s not without problems—for example, it prevents a derived class from overriding the method to return a `str`.

So I think RET501 and PLR1711 should exempt annotated functions with a return type other than `-> None`.

--- a/crates/ruff/resources/test/fixtures/flake8_return/RET501.py
+++ b/crates/ruff/resources/test/fixtures/flake8_return/RET501.py
@@ -2,3 +2,13 @@ def x(y):
     if not y:
         return
     return None  # error
+
+
+class BaseCache:
+    def get(self, key: str) -> str | None:
+        print(f"{key} not found")
+        return None
+
+    def get(self, key: str) -> None:
+        print(f"{key} not found")
+        return None

--- a/crates/ruff/resources/test/fixtures/pylint/useless_return.py
+++ b/crates/ruff/resources/test/fixtures/pylint/useless_return.py
@@ -48,3 +48,13 @@ def print_python_version():
     """This function returns None."""
     print(sys.version)
     return None  # [useless-return]
+
+
+class BaseCache:
+    def get(self, key: str) -> str | None:
+        print(f"{key} not found")
+        return None
+
+    def get(self, key: str) -> None:
+        print(f"{key} not found")
+        return None

--- a/crates/ruff/src/checkers/ast/mod.rs
+++ b/crates/ruff/src/checkers/ast/mod.rs
@@ -434,11 +434,11 @@ where
                     Rule::SuperfluousElseContinue,
                     Rule::SuperfluousElseBreak,
                 ]) {
-                    flake8_return::rules::function(self, body);
+                    flake8_return::rules::function(self, body, returns);
                 }
 
                 if self.settings.rules.enabled(Rule::UselessReturn) {
-                    pylint::rules::useless_return(self, stmt, body);
+                    pylint::rules::useless_return(self, stmt, body, returns);
                 }
 
                 if self.settings.rules.enabled(Rule::ComplexStructure) {

--- a/crates/ruff/src/checkers/ast/mod.rs
+++ b/crates/ruff/src/checkers/ast/mod.rs
@@ -434,11 +434,20 @@ where
                     Rule::SuperfluousElseContinue,
                     Rule::SuperfluousElseBreak,
                 ]) {
-                    flake8_return::rules::function(self, body, returns);
+                    flake8_return::rules::function(
+                        self,
+                        body,
+                        returns.as_ref().map(|expr| &**expr),
+                    );
                 }
 
                 if self.settings.rules.enabled(Rule::UselessReturn) {
-                    pylint::rules::useless_return(self, stmt, body, returns);
+                    pylint::rules::useless_return(
+                        self,
+                        stmt,
+                        body,
+                        returns.as_ref().map(|expr| &**expr),
+                    );
                 }
 
                 if self.settings.rules.enabled(Rule::ComplexStructure) {

--- a/crates/ruff/src/rules/flake8_annotations/helpers.rs
+++ b/crates/ruff/src/rules/flake8_annotations/helpers.rs
@@ -21,7 +21,7 @@ pub(super) fn match_function_def(stmt: &Stmt) -> (&str, &Arguments, Option<&Expr
             returns,
             body,
             ..
-        } => (name, args, returns.as_ref().map(|x| &**x), body),
+        } => (name, args, returns.as_ref().map(|expr| &**expr), body),
         _ => panic!("Found non-FunctionDef in match_name"),
     }
 }

--- a/crates/ruff/src/rules/flake8_annotations/helpers.rs
+++ b/crates/ruff/src/rules/flake8_annotations/helpers.rs
@@ -6,9 +6,7 @@ use ruff_python_ast::visibility;
 use crate::checkers::ast::Checker;
 use crate::docstrings::definition::{Definition, DefinitionKind};
 
-pub(super) fn match_function_def(
-    stmt: &Stmt,
-) -> (&str, &Arguments, &Option<Box<Expr>>, &Vec<Stmt>) {
+pub(super) fn match_function_def(stmt: &Stmt) -> (&str, &Arguments, Option<&Expr>, &Vec<Stmt>) {
     match &stmt.node {
         StmtKind::FunctionDef {
             name,
@@ -23,7 +21,7 @@ pub(super) fn match_function_def(
             returns,
             body,
             ..
-        } => (name, args, returns, body),
+        } => (name, args, returns.as_ref().map(|x| &**x), body),
         _ => panic!("Found non-FunctionDef in match_name"),
     }
 }

--- a/crates/ruff/src/rules/flake8_return/rules.rs
+++ b/crates/ruff/src/rules/flake8_return/rules.rs
@@ -4,6 +4,7 @@ use rustpython_parser::ast::{Constant, Expr, ExprKind, Location, Stmt, StmtKind}
 use ruff_diagnostics::{AlwaysAutofixableViolation, Violation};
 use ruff_diagnostics::{Diagnostic, Fix};
 use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::is_const_none;
 use ruff_python_ast::helpers::{elif_else_range, end_of_statement};
 use ruff_python_ast::types::Range;
 use ruff_python_ast::visitor::Visitor;
@@ -484,7 +485,7 @@ fn superfluous_else(checker: &mut Checker, stack: &Stack) -> bool {
 }
 
 /// Run all checks from the `flake8-return` plugin.
-pub fn function(checker: &mut Checker, body: &[Stmt]) {
+pub fn function(checker: &mut Checker, body: &[Stmt], returns: &Option<Box<Expr>>) {
     // Skip empty functions.
     if body.is_empty() {
         return;
@@ -535,7 +536,10 @@ pub fn function(checker: &mut Checker, body: &[Stmt]) {
 
     if !result_exists(&stack.returns) {
         if checker.settings.rules.enabled(Rule::UnnecessaryReturnNone) {
-            unnecessary_return_none(checker, &stack);
+            // Skip functions that have a return annotation that is not `None`.
+            if returns.is_none() || is_const_none(returns.as_ref().unwrap()) {
+                unnecessary_return_none(checker, &stack);
+            }
         }
         return;
     }

--- a/crates/ruff/src/rules/flake8_return/rules.rs
+++ b/crates/ruff/src/rules/flake8_return/rules.rs
@@ -537,7 +537,7 @@ pub fn function(checker: &mut Checker, body: &[Stmt], returns: Option<&Expr>) {
     if !result_exists(&stack.returns) {
         if checker.settings.rules.enabled(Rule::UnnecessaryReturnNone) {
             // Skip functions that have a return annotation that is not `None`.
-            if returns.is_none() || is_const_none(returns.as_ref().unwrap()) {
+            if returns.map_or(true, is_const_none) {
                 unnecessary_return_none(checker, &stack);
             }
         }

--- a/crates/ruff/src/rules/flake8_return/rules.rs
+++ b/crates/ruff/src/rules/flake8_return/rules.rs
@@ -485,7 +485,7 @@ fn superfluous_else(checker: &mut Checker, stack: &Stack) -> bool {
 }
 
 /// Run all checks from the `flake8-return` plugin.
-pub fn function(checker: &mut Checker, body: &[Stmt], returns: &Option<Box<Expr>>) {
+pub fn function(checker: &mut Checker, body: &[Stmt], returns: Option<&Expr>) {
     // Skip empty functions.
     if body.is_empty() {
         return;

--- a/crates/ruff/src/rules/pylint/rules/useless_return.rs
+++ b/crates/ruff/src/rules/pylint/rules/useless_return.rs
@@ -50,7 +50,7 @@ pub fn useless_return<'a>(
     checker: &mut Checker<'a>,
     stmt: &'a Stmt,
     body: &'a [Stmt],
-    returns: &Option<Box<Expr>>,
+    returns: Option<&'a Expr>,
 ) {
     // Skip functions that have a return annotation that is not `None`.
     if let Some(returns) = returns {

--- a/crates/ruff/src/rules/pylint/rules/useless_return.rs
+++ b/crates/ruff/src/rules/pylint/rules/useless_return.rs
@@ -1,5 +1,5 @@
 use log::error;
-use rustpython_parser::ast::{Constant, ExprKind, Stmt, StmtKind};
+use rustpython_parser::ast::{Constant, Expr, ExprKind, Stmt, StmtKind};
 
 use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic};
 use ruff_macros::{derive_message_formats, violation};
@@ -46,7 +46,19 @@ impl AlwaysAutofixableViolation for UselessReturn {
 }
 
 /// PLR1711
-pub fn useless_return<'a>(checker: &mut Checker<'a>, stmt: &'a Stmt, body: &'a [Stmt]) {
+pub fn useless_return<'a>(
+    checker: &mut Checker<'a>,
+    stmt: &'a Stmt,
+    body: &'a [Stmt],
+    returns: &Option<Box<Expr>>,
+) {
+    // Skip functions that have a return annotation that is not `None`.
+    if let Some(returns) = returns {
+        if !is_const_none(returns) {
+            return;
+        }
+    }
+
     // Skip empty functions.
     if body.is_empty() {
         return;

--- a/crates/ruff/src/rules/pylint/rules/useless_return.rs
+++ b/crates/ruff/src/rules/pylint/rules/useless_return.rs
@@ -53,10 +53,8 @@ pub fn useless_return<'a>(
     returns: Option<&'a Expr>,
 ) {
     // Skip functions that have a return annotation that is not `None`.
-    if let Some(returns) = returns {
-        if !is_const_none(returns) {
-            return;
-        }
+    if !returns.map_or(true, is_const_none) {
+        return;
     }
 
     // Skip empty functions.

